docs: toFQDNs poller upgrade impact, section reshuffle

raybejjani opened the pull request:
I've introduced a `cilium preflight` command that creates a file readable by `--tofqdns-pre-cache`. The preflight pod now runs a container with this command. Using the file is opt-in, however. The image for this container will have to change once we merge (or just before) to refer to an appropriate nightly or latest. The instructions now reflect the steps needed to upgrade poller-to-poller and then poller-to-proxy. I've left them a little sparse, figuring it is easier to add a missing bit than remove a complex paragraph. Let me know what is missing! In fact, do try out the steps. It should be a 5-minute thing with a minikube.

Note: I have a temp commit to use to the correct image and opt-into the pre-cache (so don't merge for me!)


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6886)
<!-- Reviewable:end -->

raybejjani committed:
docs: Add FQDN Poller upgrade impact & instructions

Upgrading from a poller-only deployment may see some drops until new DNS
requests are seen. This may not be viable and we can pre-seed the DNS
cache via `--tofqdns-pre-cache`.

Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani committed:
docs: Small changes to toFQDN and DNS sections

Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani committed:
docs: Move "Obtaining DNS Data" to L7 section

Signed-off-by: Ray Bejjani <ray@covalent.io>

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -448,16 +448,17 @@ provided in DNS responses are allowed by Cilium in a similar manner to IPs in
-collected from all DNS responses seen by Cilium on the node. Multiple selectors
-may be included in a single egress rule.
+IP information is captured from DNS responses per-Endpoint via a `DNS Proxy`_
+or `DNS Polling`_. A L3 `CIDR based`_ rule is generated for every ``toFQDNs``
```
```suggestion
or `DNS Polling`_. An L3 `CIDR based`_ rule is generated for every ``toFQDNs``
```

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -1032,6 +944,97 @@ allowed but connections to the returned IPs are not, as there is no L3
+  policy for each endpoint and increment the per cilium-agent policy repository
+  revision.
+
+  Polling can be enabled by the ``--tofqdns-enable-poller`` cilium-agent
```
```suggestion
  Polling may be enabled by the ``--tofqdns-enable-poller`` cilium-agent
```

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -1032,6 +944,97 @@ allowed but connections to the returned IPs are not, as there is no L3
+  revision.
+
+  Polling can be enabled by the ``--tofqdns-enable-poller`` cilium-agent
+  CLI option.
```
```suggestion
  CLI option. It is disabled by default.
```

joestringer commented on Documentation/policy/language.rst:
```diff
@@ -1032,6 +944,97 @@ allowed but connections to the returned IPs are not, as there is no L3
+
+.. _DNS Proxy:
+
+DNS Proxy
```
```suggestion
DNS Proxy (preferred)
```

joestringer replied:
I was worried this might end up putting `(preferred)` in the middle of sentences everywhere, but it appears not. I think it's useful to highlight that this approach is more robust than the other, though.

joestringer commented on Documentation/policy/language.rst:
````diff
@@ -448,16 +448,17 @@ provided in DNS responses are allowed by Cilium in a similar manner to IPs in
+rule and applies to the same endpoints. The IP information is selected for
+insertion by ``matchName`` or ``matchPattern`` rules, and is collected from all
+DNS responses seen by Cilium on the node. Multiple selectors may be included in
+a single egress rule. See `Obtaining DNS Data for use by ``toFQDNs```_ for
````
I see this new warning:

```
/srv/Documentation/policy/language.rst:451: WARNING: Unknown target name: "obtaining dns data for use by ``tofqdns``".
```

You can put a label above the relevant section, and insert a `:ref...` tag here and it will replace the `:ref:...` tag with the title of the section, then sphinx won't get confused by all the backticks here.

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -278,6 +278,35 @@ Changes that may require action
+
+.. only:: epub or latex
+
+        .. literalinclude:: ../../examples/policies/l7/dns/dns-upgrade.json
```
nit: The indentation of these is back out to the upper level, rather than indented the same as the bullet point for this item in the section about required actions.

raybejjani replied:
I think I sorted this out. I'm not sure if I understood what you meant (you mean in the rendered html, right?)

joestringer commented on Documentation/install/upgrade.rst:
```diff
@@ -278,6 +278,35 @@ Changes that may require action
+
+   In the following example egress rules to allow DNS access to kube-dns in
+   added, including a L7 DNS rule to allow interception of DNS traffic without
+   blocking any. The ``toFQDNs`` rules do not need to be modified. 
```
I had a few thoughts when reading this:
* Do I want to enable dns-poller?
* Why is it disabled by default now?
* Some of the wording is a bit too passive

I've tried to streamline the description above, I hope this is an improvement. Feel free to take/leave whichever parts of the below that you think make positive changes to the wording:

```
 * The :ref:`DNS Polling` option used by ``toFQDNs.matchName`` rules is         
   disabled by default in 1.4.x due to :ref:`limitations in the implementation  
   <DNS Polling>`. It has been replaced by :ref:`DNS Proxy` support, which must 
   be explicitly enabled via changes to the policy described below. To ease     
   upgrade, users may opt to enable the :ref:`DNS Polling` in v1.4.x by adding  
   the ``--tofqdns-enable-poller`` option to cilium-agent without changing      
   policies.                                                                    
                                                                                
   Policy rules may be prepared to use the :ref:`DNS Proxy` before an upgrade   
   to 1.4. The new policy rule fields ``toFQDNs.matchPattern`` and              
   ``toPorts.rules.dns.matchName/matchPattern`` will be ignored by older cilium 
   versions and can be safely implemented prior to an upgrade.                  
                                                                                
   The following example allows DNS access to ``kube-dns`` via the :ref:`DNS    
   Proxy` and allows all DNS requests to ``kube-dns``. For completeness,        
   ``toFQDNs`` rules are included for examples of the syntax for those L3       
   policies as well. Existing ``toFQDNs`` rules do not need to be modified.  
```

joestringer reviewed (changes requested):
Thanks for writing this up! I have a few suggestions below.

raybejjani committed:
cilium preflight command for FQDN poller upgrade

Upgrading from a poller deployment may result in interrupted connections
until DNS requests can be intercepted. We can use --tofqdns-pre-cache to
preload DNS data and avoid this initial delay.

Signed-off-by: Ray Bejjani <ray@covalent.io>

raybejjani committed:
cilium preflight container prepares tofqdn-pre-cache

We already use pre-flight to preload the cilium container image. We can
also use it to create a pre-cache file for use when upgrading cilium
from DNS Polling setups or simply to ensure some IPs are present.

Signed-off-by: Ray Bejjani <ray@covalent.io>

tgraf commented on cilium/cmd/preflight.go:
```diff
@@ -0,0 +1,138 @@
+	for name, IPs := range DNSData {
+		cache.Update(lookupTime, name, IPs, toFQDNsPreCacheTTL)
+	}
+	logrus.WithError(err).WithField("cache", cache).Debug("")
```
old?

raybejjani replied:
yup, it was!

tgraf commented on cilium/cmd/preflight.go:
```diff
@@ -0,0 +1,138 @@
+	if err != nil {
+		Fatalf("Cannot extract DNS data from local cilium-agent: %s", err)
+	}
+	logrus.WithError(err).WithField("DNSData", DNSData).Debug("")
```
old?

raybejjani replied:
yup, it was!

aanm reviewed (commented):
I'm ok with the changes but having 2 containers in the pre-flight DS might cause the pre-flight documentation to be updated as well since it will shown as the docs are marked with `READY 1/1` but in reality it will be `READY 2/2` AFAIK
